src/Common/src/System/Security/Cryptography/Asn1V2.cs

```diff
@@ -0,0 +1,5071 @@
+            }
+
+            // Slice first so that an out of bounds value triggers a CryptographicException.
+            ReadOnlyMemory<byte> contents = Slice(_data, headerLength, length.Value);
+            ReadOnlySpan<byte> contentSpan = contents.Span;
+
+            // T-REC-X.690-201508 sec 8.3.2
+            if (contents.Length > 1)
+            {
+                ushort bigEndianValue = (ushort)(contentSpan[0] << 8 | contentSpan[1]);
+                const ushort RedundancyMask = 0b1111_1111_1000_0000;
+                ushort masked = (ushort)(bigEndianValue & RedundancyMask);
+
+                // If the first 9 bits are all 0 or are all 1, the value is invalid.
+                if (masked == 0 || masked == RedundancyMask)
+                {
+                    throw new CryptographicException(SR.Cryptography_Der_Invalid_Encoding);
+                }
+            }
+
+            return contents;
+        }
+
+        public ReadOnlyMemory<byte> GetIntegerBytes() =>
+            GetIntegerBytes(Asn1Tag.Integer);
+
+        public ReadOnlyMemory<byte> GetIntegerBytes(Asn1Tag expectedTag)
+        {
+            ReadOnlyMemory<byte> contents =
+                GetIntegerContents(expectedTag, UniversalTagNumber.Integer, out int headerLength);
+
+            _data = _data.Slice(headerLength + contents.Length);
+            return contents;
+        }
+
+        private bool TryReadSignedInteger(
+            int sizeLimit,
+            Asn1Tag expectedTag,
+            UniversalTagNumber tagNumber,
+            out long value)
+        {
+            Debug.Assert(sizeLimit <= sizeof(long));
+
+            ReadOnlyMemory<byte> contents = GetIntegerContents(expectedTag, tagNumber, out int headerLength);
+
+            if (contents.Length > sizeLimit)
+            {
+                value = 0;
+                return false;
+            }
+
+            ReadOnlySpan<byte> contentSpan = contents.Span;
+
+            bool isNegative = contentSpan[0] >= 0x80;
+            long accum = isNegative ? -1 : 0;
+
+            for (int i = 0; i < contents.Length; i++)
+            {
+                accum <<= 8;
+                accum |= contentSpan[i];
+            }
+
+            _data = _data.Slice(headerLength + contents.Length);
+            value = accum;
+            return true;
+        }
+
+        private bool TryReadUnsignedInteger(
+            int sizeLimit,
+            Asn1Tag expectedTag,
+            UniversalTagNumber tagNumber,
+            out ulong value)
+        {
+            Debug.Assert(sizeLimit <= sizeof(ulong));
+
+            ReadOnlyMemory<byte> contents = GetIntegerContents(expectedTag, tagNumber, out int headerLength);
+            ReadOnlySpan<byte> contentSpan = contents.Span;
+            int contentLength = contents.Length;
+
+            bool isNegative = contentSpan[0] >= 0x80;
+
+            if (isNegative)
+            {
+                // TODO/Review: Should this be "false", an Exception, or not a scenario?
```

Comment:
False or an exception seems reasonable since it won't do what the user is asking for.